### PR TITLE
unicode_literals for bytes to unicode.

### LINKED ISSUE #11: write_document() should support unicode in addition to str
See the discussion at https://github.com/nexB/scancode-toolkit/issues/436#issuecomment-270935436.

### PR DESCRIPTION
**Pull request by rtgdk**
Fixes #11 
`from __future__ import unicode_literals ` will change bytes to unicode if run in Python2. In Python3, it won't have any affect. 
Please review it, if it's correct then I'll extend it to other files.

### CODE REVIEW

**Comment by pombredanne** [comment]
@rtgdk Thanks. But please do not pile unrelated commits in the same pull request. Can you submit a separate request for your unicode related commit?

File: spdx/writers/tagvalue.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
-
+from __future__ import unicode_literals
```
Comment (pombredanne) [inline spdx/writers/tagvalue.py]: Please keep an empty line after __future__ import

**Comment by rtgdk** [comment]
@pombredanne  I have now imported unicode_literals wherever necessary.

**Comment by rtgdk** [comment]
@pombredanne  Any review on this ? Will this work ? 
FWIW my basic testing on this worked. Since unicode_literals will only affect Python2 bytes and convert them to unicode. str of Python3 is already unicode. So we don't have to write like `u"string"`  anymore.

**Comment by pombredanne** [comment]
> Any review on this ? Will this work ?

Did you check the results on the CI tests? Are they passing everywhere on Py2 and 3?

**Comment by rtgdk** [comment]
@pombredanne  After looking through the tests failure and code, I realised I imported unicode_literals unnecessarily in some files which led to conversion of some bytes(str) to unicode where files are being opened in binary mode to read/write.( An error in python 2)
I have rectified my error. The tests also seem to pass.

**Comment by rtgdk** [comment]
@pombredanne Did you have a look at the changes?

**Comment by pombredanne** [comment]
Thanks for the update!

1. To use unicode throughout, you want to import unicode_literals everywhere. not somewhere.
2. you need to add new tests in any case.

**Comment by rtgdk** [comment]
But if I import unicode_literals everywhere then some of the current test are bound to fail. Like this `test_rdf_rdf` in `test_conversion.py` . Since the file is being opened with `open` and not with `codec.open`. Also the file is being opened in `rb` without any encoding/decoding. So it needs data in bytes and importing unicode_literals in those files will raise encoding/decoding error.
But if the file is being opened with codec.open in "w" with `encoding="utf-8"` all is working perfectly since all the strings are implicitly converted to unicode before writing.

**Comment by pombredanne** [comment]
Yes,  you need to fix the code, fix existing tests and write new tests. 
Adding the unicode_literals import is only the simple first step.

File: spdx/document.py [hunk 2]
```diff
@@ -51,7 +52,7 @@ def from_full_name(cls, full_name):
         the identifier is retrieved from it. Otherwise
         the identifier is the same as the full_name.
         """
-        if full_name in config.LICENSE_MAP.keys():
+        if full_name.encode("utf-8") in config.LICENSE_MAP.keys():
```
Comment (pombredanne) [inline spdx/document.py]: Based on my overall review comment, this should never be needed here. e.g. if you use unicode throughout, do any needed conversion at the boundaries and not here.

File: spdx/document.py [hunk 3]
```diff
@@ -40,7 +41,7 @@ def from_identifier(cls, identifier):
         the full_name is retrived from it. Otherwise
```
Comment (pombredanne) [inline spdx/document.py]: Do not hesitate to fix typos as you go: this should be "retrieved" , not "retrived"

File: spdx/parsers/loggers.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
-
+from __future__ import unicode_literals
```
Comment (pombredanne) [inline spdx/parsers/loggers.py]: Always keep two empty lines after imports

File: spdx/parsers/tagvalue.py [hunk 5]
```diff
@@ -507,7 +508,7 @@ def p_conc_license_3(self, p):
         else:
             value = p[1]
         ref_re = re.compile('LicenseRef-.+', re.UNICODE)
-        if (p[1] in config.LICENSE_MAP.keys()) or (ref_re.match(p[1]) is not None):
+        if (p[1].encode("utf-8") in config.LICENSE_MAP.keys()) or (ref_re.match(p[1]) is not None):
```
Comment (pombredanne) [inline spdx/parsers/tagvalue.py]: Do not encode here but at the boundaries

File: tests/test_document.py [hunk 6]
```diff
@@ -71,14 +72,14 @@ def test_license_list(self):
 
     def test_from_full_name(self):
         mit = License.from_full_name('MIT License')
-        assert mit.identifier == 'MIT'
+        assert mit.identifier == 'MIT'.decode("utf-8")
```
Comment (pombredanne) [inline tests/test_document.py]: Do not decode or encode here but at the boundaries

File: tests/test_document.py~ [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2014 Ahmed H. Ismail
```
Comment (pombredanne) [inline tests/test_document.py~]: do not commit editor backup files like this file

File: tests/test_document.py [hunk 8]
```diff
@@ -72,12 +72,12 @@ def test_license_list(self):
 
     def test_from_full_name(self):
         mit = License.from_full_name('MIT License')
-        assert mit.identifier == 'MIT'.decode("utf-8")
+        assert mit.identifier.decode("utf-8") == 'MIT'
```
Comment (pombredanne) [inline tests/test_document.py]: Do not decode or encode here but at the boundaries

File: tests/test_document.py [hunk 9]
```diff
@@ -72,12 +72,12 @@ def test_license_list(self):
...
         assert mit.url == 'http://spdx.org/licenses/MIT'
 
     def test_from_identifier(self):
         mit = License.from_identifier('MIT')
-        assert mit.full_name.encode("utf-8") == 'MIT License'.encode("utf-8")
+        assert mit.full_name == 'MIT License'.encode("utf-8")
```
Comment (pombredanne) [inline tests/test_document.py]: Do not decode or encode here but at the boundaries

**Comment by rtgdk** [comment]
@pombredanne I will fix all those things once I have a basic idea where these tests are failing. I know we should encode and decode at the boundaries but I just wanted a look where these implicit conversion are failing. About that commit message I know that commit messages should be clear and concise but I was just testing some changes to encoding and decoding. I plan to squash all the commits so these commit messages were just a way to push to github fast.

Apart from that, I have succesfully passed Python2 tests. But there is a new error in Python3 which never occured before in any of the test commit I did. And I don't know where I am going wrong.
Here is the code from `test_document.py` 

>  assert mit.identifier == 'MIT'

This is throwing an assertionError. i have tried decoding/encoding wherever possible but can't seem to figure out the error in Python3. **Please help me with this.**

**Comment by pombredanne** [comment]
To solve your problem, you should never have to encode or decode. Just delete any of these. Just ensure instead that whenever you read or write something in and out you only return a proper str  (which will be always unicode since you are either Py3 or using Py2 with unicode_literals.) and never bytes

**Comment by rtgdk** [comment]
@pombredanne  Yes that was my basic idea after reading about unicode changes in python2 and 3 on the web. https://nedbatchelder.com/text/unipain.html This article too said the same thing.
I did plan to remove encode/decode once I knew where the error was occuring (thats why all these commits ) and add them at the boundaries.
Maybe this was causing assertion error.
I'll rectify it as soon as possible.

**Comment by pombredanne** [comment]
@nedbat 's article is an excellent reference. Ned is also super helpful and active on the #python IRC channel on Freenode with the nedbat nick.

**Comment by pombredanne** [comment]
Though he is US east coast time.

**Comment by rtgdk** [comment]
@pombredanne  Finally after more than 40 commits, I think I have fixed the existing test. Once you review the final changes, I'll squash the commits.
I have learned a lot about unicode, their conversions, their behaviour when reading/writing in file in the last few hours. Thanks to you.
I hope you like these changes.

**Comment by rtgdk** [comment]
@pombredanne  Did you have a look at the changes ?
As you suggested, decoding/encoding at the boundaries. So I added unicode conversion when the data is being stored in the dictionary rather than opening the file with `codecs with utf8 encoding` as this was causing error with `csv` module. 
Apart from that I have added try except once while reading` '\n'.` Either this way or I can open the initial file with codecs and then remove all the decoding that being done in parsers.tagvalue.py.

**Comment by pombredanne** [comment]
@rtgdk Consider using `unicodecsv` from Pypi then: the stdlib `csv` does not support unicode. This would make the solution much cleaner

**Comment by pombredanne** [comment]
Using codecs is the way to go IMHO. Then fixing any issue that will come up

**Comment by rtgdk** [comment]
On it. Squashing all the previous commits.

**Comment by rtgdk** [comment]
@pombredanne  Tests cannot import `unicodecsv`. It's working fine in terminal. What to do ?
Also it's written in the official documentation of unicodecsv that there is no need to use `codecs` wrapper since unicodecsv expect bytestream and not unicode. So what to do? Use `open` only ?

**Comment by rtgdk** [comment]
@pombredanne  Did you have a look at the tests? What should I do ?

**Comment by rtgdk** [comment]
Travis and CI tests are failing. They cannot import unicodecsv. When I tried it out locally, it seemed to work without any additionally library download.

**Comment by pombredanne** [comment]
What happens if you clean your venv, create a clean one and then run `python setup.py install` in the vend and then run the tests with  `py.test -vvs -x`?

File: spdx/config.py~ [hunk 10]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (pombredanne) [inline spdx/config.py~]: Do NOT commit ~ files

File: spdx/config.py [hunk 11]
```diff
@@ -12,22 +12,22 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
+from __future__ import unicode_literals
 
-import csv
+import unicodecsv as csv
```
Comment (pombredanne) [inline spdx/config.py]: Do not use `as` here.

File: spdx/config.py [hunk 12]
```diff
@@ -23,15 +23,11 @@
 class TwoWayDict(dict):
 
 	def __setitem__(self, key, value):
-		try:
-			dict.__setitem__(self, unicode(key,encoding="utf-8"), unicode(value,encoding="utf-8"))
-			dict.__setitem__(self, unicode(value,encoding="utf-8"), unicode(key,encoding="utf-8"))
-		except NameError :
-			dict.__setitem__(self,key,value)
-			dict.__setitem__(self,value,key)
+		dict.__setitem__(self, key,value)
```
Comment (pombredanne) [inline spdx/config.py]: Is the spacing 4 spaces?

File: README.md [hunk 13]
```diff
@@ -1,5 +1,15 @@
-Python SPDX Library to parse, validate and create SPDX documents
-==========================
+# Python SPDX Library to parse, validate and create SPDX documents
```
Comment (pombredanne) [inline README.md]: How is this related to this PR?

File: spdx/config.py [hunk 14]
```diff
@@ -12,22 +12,22 @@
...
-        dict.__setitem__(self, key, value)
-        dict.__setitem__(self, value, key)
-
-    def __delitem__(self, key):
-        dict.__delitem__(self, self[key])
-        dict.__delitem__(self, key)
+	def __setitem__(self, key, value):
+		dict.__setitem__(self, key,value)
+		dict.__setitem__(self, value,key)
+	def __delitem__(self, key):
```
Comment (pombredanne) [inline spdx/config.py]: Why removing the empty line? And what about 4 spaces indent?

File: spdx/parsers/lexers/tagvalue.py [hunk 15]
```diff
@@ -95,7 +96,11 @@ def t_text_end(self, t):
         t.type = 'TEXT'
         t.value = t.lexer.lexdata[
             t.lexer.text_start:t.lexer.lexpos].strip()
-        t.lexer.lineno += t.value.count('\n')
+        try :
+        	unicode
```
Comment (pombredanne) [inline spdx/parsers/lexers/tagvalue.py]: What is unicode meaning here? This does not make sense.

File: spdx/parsers/lexers/tagvalue.py [hunk 16]
```diff
@@ -95,7 +96,11 @@ def t_text_end(self, t):
...
+        	t.lexer.lineno += t.value.count('\n'.encode("utf-8"))
```
Comment (pombredanne) [inline spdx/parsers/lexers/tagvalue.py]: Why are you encoding to bytes here?

**Comment by pombredanne** [comment]
Hint: add unicodecsv to setup.py's `install_requires`

File: setup.py~ [hunk 17]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
Comment (pombredanne) [inline setup.py~]: Please, do not commit editor backup files...

File: spdx/config.py [hunk 18]
```diff
@@ -32,7 +33,7 @@ def __delitem__(self, key):
 
 def load_license_list():
     FILE_NAME = os.path.join(os.path.dirname(__file__), 'spdx_licenselist.csv')
-    with open(FILE_NAME, 'r') as file_in:
+    with codecs.open(FILE_NAME, 'r', encoding="utf-8") as file_in:
```
Comment (pombredanne) [inline spdx/config.py]: I know it is not your doing, but improving the code while you touch it is a good idea.  FILE_NAME and uppercase names are used for global constants by convention and not scoped variables. Use `file_name` instead.

File: spdx/config.py~ [hunk 19]
```diff
@@ -32,8 +33,8 @@ class TwoWayDict(dict):
 
 def load_license_list():
     FILE_NAME = os.path.join(os.path.dirname(__file__), 'spdx_licenselist.csv')
-    with open(FILE_NAME, 'r') as file_in:
-        reader = csv.DictReader(file_in)
+    with codecs.open(FILE_NAME, 'r', encoding="utf-8") as file_in:
```
Comment (pombredanne) [inline spdx/config.py~]: FILE_NAME and uppercase names are used for global constants by convention and not scoped variables. Use `file_name` instead.

File: spdx/config.py (hunk 14 above)
Comment (rtgdk) [inline spdx/config.py]: I may have missed it. I use tab for indentation which is set to default 4 spaces. Should I use 2 spaces ? I think the whole code is written with 4 spaces indentation.

File: spdx/parsers/lexers/tagvalue.py (hunk 15 above)
Comment (rtgdk) [inline spdx/parsers/lexers/tagvalue.py]: It will raise an error in Python3. So the exception will be executed.

File: spdx/parsers/lexers/tagvalue.py (hunk 16 above)
Comment (rtgdk) [inline spdx/parsers/lexers/tagvalue.py]: Actually tagvalue.py in parsers has a lot of decode options, which was raising error while opening the file with `codecs`. So I plan to not change the current test ( openfile with `open` ) and change the code only once here.  
Default encoding is ascii in `open`, so I have to specifically encode it to utf-8. This is needed to pass the test I think.

File: setup.py~ (hunk 17 above)
Comment (rtgdk) [inline setup.py~]: I am really sorry. I plan to remove all those backup files and write correct commit message once I get the correct solution. These are just for fast git push to github. I'll squash all the commits once I'm done.

File: spdx/config.py (hunk 18 above)
Comment (rtgdk) [inline spdx/config.py]: Ok I'll do that.

**Comment by rtgdk** [comment]
@pombredanne I have a doubt. While using unicodecsv, the module is gonna convert bytes to unicode ( See. So it expects data to be in bytes. But in python3, data would be in str. So while decoding, it will raise error. ( See travis test : https://travis-ci.org/spdx/tools-python/jobs/215542055 ).
What should I do now. Convert file to byte stream ?

**Comment by pombredanne** [comment]
@rtgdk If you have some issues with unicodecsv I suggest filing a bug with them. The best way would be to write a test that shows the issue first.

**Comment by pombredanne** [comment]
@rtgdk note also that you will need to rebase your PR on the latest master

**Comment by pombredanne** [comment]
@rtgdk It sounds like you merged and did not rebase? or something else?

File: README.md [hunk 20]
```diff
@@ -1,5 +1,15 @@
...
+
+| Linux | macOS | Windows |
```
Comment (pombredanne) [inline README.md]: How come these files are part of your PR? I am a tad confused.

**Comment by rtgdk** [comment]
@pombredanne  Sorry, for not getting back to this earlier. My exams were going on.
I have currently written a simple test to show `unicodecsv` error in Python3.
I will rebase and merge with the master branch as soon as I make the changes in `config.py` to read the file using `csv`.

**Comment by rtgdk** [comment]
@pombredanne  I have made the `unicodecsv` test. It is working in python2 but not in Python3.( That is why Travis test failed)

Also instead of `unicodecsv` I am using `csv` module for compatibility with both 2 and 3 version.

**Comment by pombredanne** [comment]
@rtgdk Thanks. Let me check that after the week end!

**Comment by rtgdk** [comment]
@pombredanne  Did you have a look at the changes ?

**Comment by pombredanne** [comment]
@rtgdk I had looked at you changes... but did you check all the comments I made while reviewing the code?

File: README.md (hunk 13 above)
Comment (pombredanne) [inline README.md]: ?

File: README.md (hunk 20 above)
Comment (pombredanne) [inline README.md]: ?

File: spdx/config.py [hunk 21]
```diff
@@ -21,18 +21,18 @@
...
 class TwoWayDict(dict):
 
-    def __setitem__(self, key, value):
-        dict.__setitem__(self, key, value)
-        dict.__setitem__(self, value, key)
-
-    def __delitem__(self, key):
-        dict.__delitem__(self, self[key])
-        dict.__delitem__(self, key)
+	def __setitem__(self, key, value):
```
Comment (pombredanne) [inline spdx/config.py]: why did you change the indenting here?

File: spdx/parsers/lexers/tagvalue.py (hunk 16 above)
Comment (pombredanne) [inline spdx/parsers/lexers/tagvalue.py]: If this is needed to pass the test, then may be the tests need to be updated instead?

File: spdx/parsers/loggers.py (hunk 4 above)
Comment (pombredanne) [inline spdx/parsers/loggers.py]: ?

File: tests/test_unicodecsv.py [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+import unittest
+
+import unicodecsv
+
+class TestUnicodeCsv(unittest.TestCase):
+
+    def test_unicode(self):
+    	dct=dict()
+    	with open('data/sample.csv', 'r') as file_in:
+		    reader = unicodecsv.DictReader(file_in)
```
Comment (pombredanne) [inline tests/test_unicodecsv.py]: Please use 4 spaces for indent, not 8

File: tests/test_unicodecsv.py [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import unittest
+
+import unicodecsv
+
+class TestUnicodeCsv(unittest.TestCase):
+
+    def test_unicode(self):
```
Comment (pombredanne) [inline tests/test_unicodecsv.py]: I am not sure we care for a test of unicodecsv. What we need is a test of the support of unicode in thsi spdx library instead. For inatnce, what if we pass unicode content? etc.

**Comment by pombredanne** [comment]
@rtgdk Thanks. You still have some tests that are failing though.

**Comment by rtgdk** [comment]
@pombredanne  Sorry for not getting back to this earlier. My exams were going on.
The tests are failing for Python3 as the key-value pair is in unicode ( `unicode_literals `). But passing the key as byte in Python3 will create `Key Error `as no such key is present in the dictionary.
I'll just change it to `assert not` in case of Python3.

**Comment by pombredanne** [comment]
Hum ok. but I cannot fathom why there would be a different behavior there.... Also did you check the other comments I made on the PR? No rush if you have exams of course :)

**Comment by rtgdk** [comment]
screenshot from 2017-05-02 20-22-18
This is why Python3 shows error. Since we are using unicode_literals, all the data is stored in unicode. So accessing them in bytes in not possible in Python3.

screenshot from 2017-05-02 20-23-13 
And this is case in Python2

**Comment by rtgdk** [comment]
I have now left two empty lines after imports.
README.md is still need to be change. But I can't seem to figure out why this file is being shown in the changes. I have taken the exact same file from spdx:master. You can see in the above screenshot that there is a field for Linux,macOS and Windows. But there is no such field in the original file. So I don't know what to do.

**Comment by pombredanne** [comment]
@rtgdk I will review and merge, and eventually cherry picking the commits as needed. Thanks!

**Comment by rtgdk** [comment]
@pombredanne  What about this PR ? Any changes needed ?

**Comment by pombredanne** [comment]
I just need to review it .... sorry for taking some time

**Comment by pombredanne** [comment]
Actually this has all been merged in alright after a rebase some while ago.... with https://github.com/spdx/tools-python/commit/dce58f5d30c792c3d582b431accfb402f175dd83  so closing
